### PR TITLE
Disable non-neutral owners for player spawns in map editor

### LINKED ISSUE #18092: Index out of bounds while loading skirmish custom map
## Issue Summary

The skirmish game crashes on loading my own custom map with a index out of bounds exception.
(Cnc)

## Reproduction

Tested with "release-20200503"

1. Select Skirmish mode
2. Select the map "Rumble in the jungle 2"
3. Choose one cabal bot player on 2nd slot
4. Select the start position "A" for the player
5. Select the start position "G" for the bot
6. Start the game -> the game crashes

## Expected behavior

The game should just start, or the map editor should tell me if something is wrong with the map.
I have chosen 8 Multiplayer positions in the map, I thought this should also work for the skirmish mode, am I wrong?

## Screenshots / Screen recordings / Replays

[rumbleInJungle2.zip](https://github.com/OpenRA/OpenRA/files/4618608/rumbleInJungle2.zip)
[exception-2020-05-12T194406Z.log](https://github.com/OpenRA/OpenRA/files/4618610/exception-2020-05-12T194406Z.log)

### PR DESCRIPTION
**Pull request by markelz0r**
Fixes #18092 

In my opinion it makes sense to remove other actors from the owners dropdown in map editor since player spawns assigned to anything other than neutral will cause crashes when spawns are unassigned

I've also added a check if we are trying to get a spawn and raise an exception if we do, however this approach is far from ideal and will still crash the entire game. I would really appreciate it If anyone could give me a lead on how could we recover from this state.

### CODE REVIEW

**Comment by pchote** [comment]
>  I would really appreciate it If anyone could give me a lead on how could we recover from this state.

The lobby does some basic validation for maps that define invalid rules, and prevents games from being started on those maps.

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs [hunk 1]
```diff
@@ -252,7 +252,17 @@ public override void Tick()
...
+						// Get if this trait requires a specific owner
+						var specificOwnerInfo = actor.Info.TraitInfoOrDefault<RequiresSpecificOwnersInfo>();
+						var owners = editorActorLayer.Players.Players.Values;
+						if (specificOwnerInfo != null)
+						{
+							owners = editorActorLayer.Players.Players.Values.Where(x =>
+								specificOwnerInfo.ValidOwnerNames.Any(y => x.Name == y))
+								.ToDictionary(player => player.Name)
+								.Values;
+						}
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs]: We should disable the dropdown too if `owners.Count == 1`.

**Comment by markelz0r** [comment]
> The lobby does some basic validation for maps that define invalid rules, and prevents games from being started on those maps.

So I've had a look into `UpdateCurrentMap()` in `LobbyLogic` and tried a few things but it seems like we cannot verify the ownership of player spawns before we create the actual World where the game would crash.  In lobby all we get is the list of spawns from yaml which is not that much useful

**Comment by pchote** [comment]
That issue is related to #17895.  It seems that there is no easy way to validate additional state in the lobby, so IMO we should just forget about this here so that the editor fix can be merged.

**Comment by markelz0r** [comment]
Sounds fair. I've added an exception there which should be a bit more descriptive when someone comes across this issue again

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs [hunk 2]
```diff
@@ -234,6 +234,25 @@ public override void Tick()
...
+					if (specificOwnerInfo != null)
+					{
+						owners = editorActorLayer.Players.Players.Values.Where(x =>
+								specificOwnerInfo.ValidOwnerNames.Any(y => x.Name == y))
+							.ToDictionary(player => player.Name)
+							.Values;
+						if (owners.Count == 1)
+						{
+							ownerDropdown.IsDisabled = () => true;
+						}
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs:249]: Single curly braces must be omitted according to https://github.com/OpenRA/OpenRA/wiki/Coding-Standard

**Comment by Mailaender** [comment]
This seems to need a rebase to get rid of the merge commit.

**Comment by abcdefg30** [comment]
No activity since a while, so closing to clean the PR queue up.

**Pull request closed without merging**